python/ray/dag/dag_node.py

```diff
@@ -380,9 +386,20 @@ def traverse_and_apply(self, fn: "Callable[[DAGNode], T]"):
         """
         visited = set()
         queue = [self]
+        num_output_nodes = 0
+
         while queue:
             node = queue.pop(0)
             if node not in visited:
+                if node.is_output_node:
+                    num_output_nodes += 1
+                    # Validate whether there are multiple nodes that call
+                    # `experimental_compile`.
+                    if num_output_nodes > 1:
+                        raise ValueError(
+                            "The DAG was compiled more than once, so some output "
+                            "nodes became leaf nodes."
```

Comment:
We can print the node for easier debugging.
Print the node at the very end of the message for better formatting (since it has multiple lines).

Reply:
Updated: https://github.com/ray-project/ray/pull/47431/commits/c31286deac9f2a6d8ff996e049675685d76d479f